### PR TITLE
Read

### PR DESCRIPTION
**Pull request by khou2020**
Read functionality for both f and g case.
Writing data read if running both case.

### CODE REVIEW

**Comment by wkliao** [comment]
Please update README.md for this new feature.
Also add test commands into .travis.yml so Travis CI can run and test.

File: e3sm_io.c [hunk 1]
```diff
@@ -84,14 +85,19 @@ int main(int argc, char** argv)
     run_g_case = 0;
 
     /* command-line arguments */
-    while ((i = getopt(argc, argv, "hkvdnmto:r:")) != EOF)
+    while ((i = getopt(argc, argv, "hkvdnmto:r:WR:")) != EOF)
```
[the remark this answers is not in the document]
Comment (khou2020) [reply e3sm_io.c]: -t is used by "Write 2D variables followed by 3D variables"

We need to replace it as well before -t can be used.

[the remark this answers is not in the document]
Comment (wkliao) [reply e3sm_io.c]: I suggest the followings
```
         [-w] perform writes
         [-r] perform reads
         [-s] Write 2D variables followed by 3D variables
         [-t num] Number of records (default 1)
         [-o output_dir] Output directory name (default ./)
```
As for the input, consider
```
         [-i input_dir] Input directory name (default ./)
```
or
```
         [-i input_file] Input file name
```

**Comment by wkliao** [comment]
I think input dir is not necessary. It can simply be an input file.
Also, when borrowing write subroutines to implement read, source code comments must be updated accordingly.

**Comment by wkliao** [comment]
also, travis ci failed with this PR.

**Comment by khou2020** [comment]
There are 2 files in f case. the file name is hardcoded as in the write only version. We need 2 separate options if we use file name instead of directory.

**Comment by wkliao** [comment]
I suggest we can adopt the approach used by the FLASH-IO benchmark, i.e. input file name prefix.
This way, we only need one command-line option for input file.

Or maybe we can just have one read case, focusing on the history 0 file (large-size one).

**Comment by khou2020** [comment]
Changed dir to prefix.
Added option to run only h0 or h1.

**Comment by wkliao** [comment]
Check the README of FLASH-IO benchmark program in PnetCDF for what I meant by "prefix".

**Comment by khou2020** [comment]
I don't see any difference. Prefix is concated directly with hard coded file name to form the final file path.

$ ./e3sm_io -t 1 -o test/flash_io_test_ -k -f 0 datasets/f_case_866x72_16p.nc
Total number of MPI processes      = 1
Input decomposition file           = datasets/f_case_866x72_16p.nc
Number of decompositions           = 3
Input file directory               =
Output file directory              = test/flash_io_test_
Variable dimensions (C order)      = 72 x 866
Write number of records (time dim) = 1
Using noncontiguous write buffer   = no
PnetCDF vard API requires internal and external data types match, skip

==== benchmarking F case writing using varn API ========================
Variable written order: same as variables are defined

History output file                = f_case_h0_varn.nc
MAX heap memory allocated by PnetCDF internally is 34.23 MiB
Total number of variables          = 408
Total write amount                 = 16.14 MiB = 0.02 GiB
Total number of requests           = 408
Max number of requests             = 408
Max Time of open + metadata define = 0.0239 sec
Max Time of I/O preparing          = 0.0596 sec
Max Time of ncmpi_iput_varn        = 0.1157 sec
Max Time of ncmpi_wait_all         = 0.0672 sec
Max Time of close                  = 0.0048 sec
Max Time of TOTAL                  = 0.2743 sec
I/O bandwidth (open-to-close)      = 58.8279 MiB/sec
I/O bandwidth (write-only)         = 238.9828 MiB/sec
-----------------------------------------------------------
$ ls test
flash_io_test_f_case_h0_varn.nc

**Comment by wkliao** [comment]
They mean differently. One is folder name and the other is file prefix name.
Source codes and comments need to reflect this intent.

**Comment by khou2020** [comment]
Changed variable name and comment

**Comment by khou2020** [comment]
Synced with latest master branch

**Pull request closed without merging**